habitat-lab/habitat/tasks/nav/nav.py

```diff
@@ -1209,39 +1270,95 @@ def step(
 
-        # Convert from [-1, 1] to [0, 1] range
-        linear_velocity = (linear_velocity + 1.0) / 2.0
-        angular_velocity = (angular_velocity + 1.0) / 2.0
+        # Apply camera action and get next observation
+        agent_state_result = self._apply_camera_pitch_velocity_action(
+            camera_pitch_ang_vel_processed,
+            time_step=time_step,
+        )
 
-        # Scale actions
-        linear_velocity = self.min_lin_vel + linear_velocity * (
-            self.max_lin_vel - self.min_lin_vel
+        # Apply action and get next observation
+        agent_state_result = self._apply_velocity_action(
+            lin_vel_processed,
+            ang_vel_processed,
+            time_step=time_step,
         )
-        angular_velocity = self.min_ang_vel + angular_velocity * (
-            self.max_ang_vel - self.min_ang_vel
+
+        return self._get_agent_observation(agent_state_result)
+
+    def _preprocess_action(
+        self, linear_velocity, angular_velocity, camera_pitch_angular_velocity
+    ):
+        """Perform scaling and clamping of input"""
+        if self._enable_scale_convert:
+            linear_velocity = self._scale_inputs(
+                linear_velocity,
+                [-1, 1],
+                self._lin_vel_range,
+            )
+            angular_velocity = self._scale_inputs(
+                angular_velocity,
+                [-1, 1],
+                self._ang_vel_range,
+            )
+            camera_pitch_angular_velocity = self._scale_inputs(
+                camera_pitch_angular_velocity,
+                [-1, 1],
+                self._ang_vel_range,
+            )
+
+        linear_velocity_clamped = np.clip(
+            linear_velocity,
+            self._lin_vel_range[0],
+            self._lin_vel_range[1],
+        )
+        angular_velocity_clamped = np.clip(
+            angular_velocity,
+            self._ang_vel_range[0],
+            self._ang_vel_range[1],
+        )
+        camera_pitch_angular_velocity_clamped = np.clip(
+            camera_pitch_angular_velocity,
+            self._ang_vel_range_camera_pitch[0],
+            self._ang_vel_range_camera_pitch[1],
         )
 
-        # Stop is called if both linear/angular speed are below their threshold
-        if (
-            abs(linear_velocity) < self.min_abs_lin_speed
-            and abs(angular_velocity) < self.min_abs_ang_speed
-        ):
-            task.is_stop_called = True  # type: ignore
-            return self._sim.get_observations_at(position=None, rotation=None)
+        return (
+            linear_velocity_clamped,
+            angular_velocity_clamped,
+            camera_pitch_angular_velocity_clamped,
+        )
+
+    def _apply_velocity_action(
+        self,
+        linear_velocity: float,
+        angular_velocity: float,
+        time_step: Optional[float] = None,
+    ):
+        """
+        Apply velocity command to simulation, step simulation, and return agent observation
```

Comment:
Add return typing.
It looks like this returns an agent_state and not an observation